src/app/core/guards/checkout-guard/checkout.guard.ts

```diff
@@ -0,0 +1,46 @@
+import { Injectable } from '@angular/core';
+import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot, UrlTree } from '@angular/router';
+import { Observable, of } from 'rxjs';
+import { Store } from '@ngrx/store';
+import { getLoggedInUser, UserState } from '../../../features/user/reducer/user.reducer';
+import { map, switchMap } from 'rxjs/operators';
+import { ProductsService } from '../../services/products/products.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CheckoutGuard implements CanActivate {
+
+  constructor(private userStore: Store<UserState>, private router: Router, private productsService: ProductsService) {
+  }
+
+  canActivate(
+    next: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot): Observable<boolean | UrlTree> | Promise<boolean | UrlTree> | boolean | UrlTree {
+    return this.productsService.getProductById(next.params.id).pipe(
+      switchMap((product) => {
+        if (product) {
+          return this.userStore.select(getLoggedInUser).pipe(
+            switchMap((loggedInUser) => {
+              if (loggedInUser) {
+                return this.productsService.getProductsByUserId(loggedInUser.id).pipe(
+                  map((userProducts) => {
+                    if (userProducts.find((product) => product.id === next.params.id)) {
```

Comment:
What are you checking here?
Is it checking whether the user is trying to lend its own product?

Reply:
First I check if the product with the given id really exist. If it is exist, yes, I check if the use try to lend his own product.

Reply:
Got it